### PR TITLE
Reimplementation of microcontroller API

### PR DESCRIPTION
**Pull request by whymarrh**
This PR is a reimplementation of the Node.js microcontrollers SDK (refs LakeMaps/microcontrollers#2).

This implementation only contains the [Wireless Communication Microcontroller API](https://github.com/LakeMaps/microcontrollers/wiki/Wireless-Communication-Microcontroller-API) where that's all that is pertinent to Hangashore.

The Implemented commands:

- [x] [`0x00`](https://github.com/LakeMaps/microcontrollers/wiki/Wireless-Communication-Microcontroller-API#command-0x00--reset-wireless-module)
- [x] [`0x03`](https://github.com/LakeMaps/microcontrollers/wiki/Wireless-Communication-Microcontroller-API#command-0x03--wireless-receive)
- [x] [`0x04`](https://github.com/LakeMaps/microcontrollers/wiki/Wireless-Communication-Microcontroller-API#command-0x04--wireless-transmit)

See also: LakeMaps/boat#19

### CODE REVIEW

**Comment by whymarrh** [comment]
refs: LakeMaps/microcontrollers#9, #18

File: src/Hangashore/lib/microcontrollers/Message.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import {crc16} from './Checksum';
+
+export class Message {
+    static from(buffer: Buffer) {
+        return new Message(buffer[1], buffer.slice(2, buffer.length - 2));
+    }
+
+    constructor(readonly command: number, readonly payload: Buffer) {
+        // ???
```
Comment (arandell93) [inline src/Hangashore/lib/microcontrollers/Message.ts:9]: Why is this section empty?

**Review by arandell93 — COMMENTED** [review]
Not gonna lie, I don't understand any of this.

File: src/Hangashore/lib/microcontrollers/Message.ts (hunk 1 above)
Comment (whymarrh) [reply src/Hangashore/lib/microcontrollers/Message.ts:9]: This constructor uses [Parameter properties](https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties), so `command` and `payload` become properties on the object

**Comment by whymarrh** [comment]
Is there something in particular I can comment on or clarify?

**Comment by arandell93** [comment]
> Is there something in particular I can comment on or clarify?

I lack a fundamental understanding of classes/constructors/objects etc. Essentially I just can't follow the structure of it because I don't have the background in software development. I'm sure you would be able to explain it to me in the next meeting.

My comment was meant more of a fulfillment of the review you requested, just without any meaningful notes from said review.

**Comment by whymarrh** [comment]
Fair enough, we can certainly go over it a later point. That said, if you have a chance, I still think that there are parts of this that will benefit from review. In particular, the tests for the sending/receiving can be double-checked for correctness along with the other parts relating to the protocol. I'll be adding the implementations of `0x00` and `0x03` in a bit, and they'll have tests too that could use a look over.

File: src/Hangashore/lib/microcontrollers/WirelessLinkMicrocontroller.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    private readonly _recv: Recv;
+
+    private readonly _send: Send;
+
+    constructor(recv: Recv, send: Send) {
+        this._recv = recv;
+        this._send = send;
+    }
+
+    reset() {
```
Comment (whymarrh) [inline src/Hangashore/lib/microcontrollers/WirelessLinkMicrocontroller.ts:43]: Does this have a return type?

**Pull request merged**